### ISSUE TITLE
Latest 3.10.y branch fails to compile with 'kuser_cmpxchg64_fixup' error

### ISSUE THREAD

**Issue by srcshelton** [raspberrypi/linux#356]
With the latest update to 3.10.6:

``` make
  LINK    vmlinux - due to target missing
  LD      vmlinux.o
  MODPOST vmlinux.o - due to vmlinux.o not in $(targets)
  GEN     .version
  CHK     include/generated/compile.h
  UPD     include/generated/compile.h
  CC      init/version.o - due to: include/generated/compile.h
  LD      init/built-in.o - due to: init/version.o
arch/arm/kernel/built-in.o: In function `__dabt_usr':
:(.text+0x470): undefined reference to `kuser_cmpxchg64_fixup'
arch/arm/kernel/built-in.o: In function `__irq_usr':
:(.text+0x4b0): undefined reference to `kuser_cmpxchg64_fixup'
make[3]: *** [vmlinux] Error 1
make[2]: *** [sub-make] Error 2
make[1]: *** [vmlinux] Error 2
make: *** [sub-make] Error 2
```

The function appears to be defined in `arch/arm/kernel/entry-armv.S` - and it looks as if this problem will only occur if CONFIG_KUSER_HELPERS is unset.

kuser_cmpxchg64_fixup is called `#if !defined(CONFIG_CPU_32v6K) && !defined(CONFIG_NEEDS_SYSCALL_FOR_CMPXCHG) && defined(CONFIG_MMU)` but is only defined `#ifdef CONFIG_KUSER_HELPERS && !defined(CONFIG_SMP) && defined(CONFIG_MMU)`.

However, CONFIG_KUSER_HELPERS appears to be a potential security vulnerability, and should only affect user-space code in any case.

It looks to me as if CONFIG_KUSER_HELPERS should be able to be safely disabled, but there's a missing define here preventing this...

**Comment by popcornmix** [comment]
I assume this is an upstream regression from 3.10.5 to 3.10.6?
If you can identify the upstream commit that broke this and report it to linux kernel mailing list that would be helpful.
If setting CONFIG_KUSER_HELPERS avoids the problem then that sounds like the workaround for now.

**Comment by hexameron** [comment]
setting CONFIG_CPU_32v6K might help also.

**Comment by srcshelton** [comment]
Hmm - it looks as if there are potentially other problems with 3.10.6.  I have a .config which has worked without fail for 3.10.x up to 3.10.4, but trying to boot 3.10.6 (with CONFIG_KUSER_HELPERS) results in:

```
[    3.663056] BCM2708FB: allocated DMA memory 5e930000
[    3.668057] BCM2708FB: allocated DMA channel 0 @ f2007000
[    3.701594] Console: switching to colour frame buffer device 201x39
[    3.717499] vchiq: vchiq_init_state: slot_zero = 0xdf804000, is_master = 0
[    3.725636] usbcore: registered new interface driver smsc95xx
[    3.731789] dwc_otg: version 3.00a 10-AUG-2012 (platform bus)
[    3.937863] Core Release: 2.80a
[    3.943816] Setting default values for core params
[    3.951354] Finished setting default values for core params
[    4.159665] Using Buffer DMA mode
[    4.165661] Periodic Transfer Interrupt Enhancement - disabled
[    4.174159] Multiprocessor Interrupt Enhancement - disabled
[    4.182415] OTG VER PARAM: 0, OTG VER FLAG: 0
[    4.189445] Dedicated Tx FIFOs mode
[    4.195767] Unable to handle kernel paging request at virtual address ffff1224
[    4.205830] pgd = c0004000
[    4.211277] [ffff1224] *pgd=1effd821, *pte=1efff0cf, *ppte=1efff21e
[    4.220410] Internal error: Oops: 81f [#1] PREEMPT ARM
[    4.228276] Modules linked in:
[    4.234051] CPU: 0 PID: 1 Comm: swapper Not tainted 3.10.6-srcs #2
[    4.242948] task: de806c00 ti: de83a000 task.ti: de83a000
[    4.251120] PC is at memcpy+0xb4/0x330
[    4.257559] LR is at 0xe1a0f008
[    4.263355] pc : [<c025c9b4>]    lr : [<e1a0f008>]    psr: 00000013
[    4.263355] sp : de83bde4  ip : 0000001c  fp : 00000000
[    4.280190] r10: c05b6f48  r9 : 00000000  r8 : ffffffff
[    4.288142] r7 : 00000000  r6 : c0591108  r5 : 00001228  r4 : 00001224
[    4.297437] r3 : c00113b8  r2 : ffffffe4  r1 : c00113bc  r0 : ffff1224
[    4.306748] Flags: nzcv  IRQs on  FIQs on  Mode SVC_32  ISA ARM  Segment kernel
[    4.316888] Control: 00c5387d  Table: 1d244008  DAC: 00000017
[    4.325422] Process swapper (pid: 1, stack limit = 0xde83a1b0)
[    4.334044] Stack: (0xde83bde4 to 0xde83c000)
[    4.341176] bde0:          00001228 c0591108 00000000 ffffffff ffff1224 00001224 c001126c
[    4.352251] be00: c058f8ac dd24d260 c05910f8 c030a4e0 c02fd888 00000020 dd24e780 c0598288
[    4.363358] be20: 00000000 00000080 de81d000 c0069ba0 de903800 c060dca4 dd24d260 c05910f8
[    4.374499] be40: de903830 c05b7028 c04381e8 c02fe48c c05b7368 de81d000 00000000 00000000
[    4.385684] be60: dd24d260 c05910f8 00000000 c0591108 c05b7028 c02f18e8 c04fee31 dd24d260
[    4.396872] be80: c0591108 c059113c c05b707c c05b707c 000000b4 c0583bb0 c056446c c02adea0
[    4.408032] bea0: c02ade8c c02acd88 00000000 c0591108 c059113c c05b707c 00000000 c02acf14
[    4.419226] bec0: c05b707c de83bed0 c02aceb4 c02ab720 de820f4c de838a50 c05b707c c05b707c
[    4.430436] bee0: dd24c120 c05b5800 00000000 c02ac520 c04370d9 c04370da c05b707c c057db34
[    4.441648] bf00: 00000000 c05c0900 000000b4 c02ad414 00000000 c05910f0 c057db34 00000000
[    4.452907] bf20: c05c0900 000000b4 c056446c c0576be0 de83a000 c0576b90 c057db34 00000000
[    4.464148] bf40: c05c0900 c0008528 00000006 00000006 00000000 00000006 c057db30 00000006
[    4.475355] bf60: c057db34 c057db14 c05c0900 000000b4 c0583bb0 c0564b6c 00000006 00000006
[    4.486596] bf80: c056446c 00000000 00000000 c03f5c10 00000000 00000000 00000000 00000000
[    4.497860] bfa0: 00000000 c03f5c18 00000000 c000d858 00000000 00000000 00000000 00000000
[    4.509060] bfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    4.520250] bfe0: 00000000 00000000 00000000 00000000 00000013 00000000 c0c0c0c0 c0c0c0c0
[    4.531397] Code: e4805004 e4806004 e4807004 e4808004 (e480e004) 
[    4.540642] ---[ end trace f51c08465fc23a13 ]---
[    4.548306] Kernel panic - not syncing: Attempted to kill init! exitcode=0x0000000b
[    4.548306] 
```

:(

**Comment by ajs124** [comment]
Another arch user and me are/were getting the same error (or at least something pretty similar) after upgrading from 3.10.4 to 3.10.6 (my ["screenshot"](http://i.imgur.com/M8Njfzj.jpg)). I downgraded to 3.6.11 for now.

**Comment by P33M** [comment]
Here's the smoking gun: Someone's been playing with my vectors page

https://lwn.net/Articles/563170/

```
Russell King (10):
      ARM: poison the vectors page
      ARM: poison memory between kuser helpers
      ARM: move vector stubs
      ARM: use linker magic for vectors and vector stubs
      ARM: update FIQ support for relocation of vectors
      ARM: allow kuser helpers to be removed from the vector page
      ARM: move signal handlers into a vdso-like page
      ARM: make vectors page inaccessible from userspace
      ARM: fix a cockup in 48be69a02 (ARM: move signal handlers into a vdso-like page)
      ARM: fix nommu builds with 48be69a02 (ARM: move signal handlers into a vdso-like page)
```

**Comment by P33M** [comment]
Regression was introduced from mainline commit http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=e39e3f3ebfef03450cf7bfa7a974a8c61f7980c8

Which has been pulled into 3.10.6 stable in commit http://git.kernel.org/cgit/linux/kernel/git/stable/linux-stable.git/commit/?h=linux-3.10.y&id=6904e468bb92a726098a2dfcf792463e11053582

The fix is now in linux-next - but it ALSO needs to be pulled down to 3.10.

Edit: This is for the panic on boot, not the compile failure.

**Comment by popcornmix** [comment]
Thanks. Rebased to 3.10.7 and cherry-picked:
http://git.kernel.org/cgit/linux/kernel/git/next/linux-next.git/commit/arch/arm/kernel/fiq.c?id=2ba85e7af4c639d933c9a87a6d7363f2983d5ada

**Comment by srcshelton** [comment]
Just a quick note to say that with 3.10.7, the compilation failure still exists if `CONFIG_KUSER_HELPERS` is not set.

**Comment by srcshelton** [comment]
In the rpi-3.11.y branch, compilation succeeds if CONFIG_KUSER_HELPERS is not set.

**Comment by popcornmix** [comment]
@srcshelton 
Interesting - I've not changed anything, so probably fixed upstream.

**Comment by szarpaj** [comment]
Why this is closed? The bug is still in 3.10.y

`arch/arm/kernel/built-in.o: In function `__dabt_usr':
:(.text+0x470): undefined reference to `kuser_cmpxchg64_fixup'
arch/arm/kernel/built-in.o: In function `__irq_usr':
:(.text+0x4b0): undefined reference to `kuser_cmpxchg64_fixup'
/home/tmp/raspi/3.10.y/Makefile:781: polecenia dla obiektu 'vmlinux' nie powiodły się
make[1]: *** [vmlinux] Błąd 1
Makefile:130: polecenia dla obiektu 'sub-make' nie powiodły się
make: *** [sub-make] Błąd 2`
